### PR TITLE
WX-1784 Compute cost from metadata

### PR DESCRIPTION
**Pull request by jgainerdewar**
### Description

 * Removed `includeTaskBreakdown` and `includeSubworkflowBreakdown` options to keep things simple for now - these aren't needed for our current work.
 * Updated stub code to actually compute workflow cost from metadata events. Note that nothing is currently *creating* the metadata events, that's in progress in another branch. For my testing, I hard-coded some extra metadata entries for each task in `StandardCachingActorHelper.startMetadataKeyValues`. 
 * Added subworkflow support

#### `CHANGELOG.md`
 - [ ] I updated `CHANGELOG.md` in this PR
 - [x] I assert that this change shouldn't be included in `CHANGELOG.md` because it doesn't impact community users

#### Terra Release Notes
 - [ ] I added a suggested release notes entry in this Jira ticket
 - [x] I assert that this change doesn't need Jira release notes because it doesn't impact Terra users

### CODE REVIEW

File: services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala [hunk 1]
```diff
@@ -269,15 +269,6 @@ trait MetadataDatabaseAccess {
...
 
-  def queryCost(id: WorkflowId, timeout: Duration)(implicit
-    ec: ExecutionContext
-  ): Future[Seq[MetadataEvent]] = {
-    val keys = List("taskStartTime", "taskEndTime", "vmCostPerHour")
-    metadataDatabaseInterface
-      .queryMetadataEntryWithKeyConstraints(id.toString, keys, List.empty, CallOrWorkflowQuery, timeout)
-      .map(metadataToMetadataEvents(id))
-  }
-
```
Comment (jgainerdewar) [inline services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala:280]: Dead code leftover from previous work in this area.

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala [hunk 2]
```diff
@@ -488,6 +515,134 @@ class MetadataBuilderActor(readMetadataWorkerMaker: () => Props,
...
+          val rawCostPerHour = extractFromJsAs[JsNumber](jsCall, CallMetadataKeys.VmCostPerHour).map(_.value)
+          val costPerHourErrorOr = rawCostPerHour match {
+            case Some(c: BigDecimal) if c >= 0 => c.validNel
+            // A costPerHour < 0 indicates an error
+            case Some(_: BigDecimal) =>
+              val index =
+                extractFromJsAs[JsNumber](jsCall, "shardIndex").map(_.value.toString).getOrElse("-1")
+              val attempt =
+                extractFromJsAs[JsNumber](jsCall, "attempt").map(_.value.toString).getOrElse("1")
+              s"Couldn't find valid vmCostPerHour for ${List(callName, index, attempt).mkString(".")}".invalidNel
```
Comment (jgainerdewar) [inline services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:578]: Note that this error message contains the call name relative to the current workflow - not fully-qualified. To make it a FQCN, we would need to modify the call name each time we get errors from a subworkflow, to prepend the "current" workflow name. That feels gross and so I haven't done it (yet?).

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala [hunk 3]
```diff
@@ -488,6 +515,134 @@ class MetadataBuilderActor(readMetadataWorkerMaker: () => Props,
...
+          }
+
+          for {
+            start <- startTimeErrorOr
+            end <- endTimeErrorOr
+            costPerHour <- costPerHourErrorOr
+            vmRuntimeInMillis = start.until(end, ChronoUnit.MILLIS).toDouble
+          } yield (vmRuntimeInMillis / (1000 * 60 * 60)) * costPerHour
+        } getOrElse (BigDecimal(0).validNel)
+      // TODO is millis the right thing to use here? Should it be seconds?
```
Comment (THWiseman) [inline services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:589]: millis seems fine to me. We're not going to run out, and we can (maybe) get a (slightly) more accurate estimation. FWIW the cost per hour already has many decimal places.

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala [hunk 4]
```diff
@@ -488,6 +515,134 @@ class MetadataBuilderActor(readMetadataWorkerMaker: () => Props,
...
+    val metadataEvents = MetadataBuilderActor.parse(groupEvents(eventsList), expandedValues)
+
+    // Walk the structured metadata to attempt to compute cost for each call in this workflow
+    val callCostsWithErrors: List[ErrorOr[BigDecimal]] = for {
+      wfObj <- extractFromJsAs[JsObject](metadataEvents, id.toString).toList
+      callsObj <- extractFromJsAs[JsObject](wfObj, "calls").toList
+      singleCallName <- callsObj.fields.keys.toList
+      singleCallAttemptList <- extractFromJsAs[JsArray](callsObj, singleCallName).toList
+      singleCallAttempt <- singleCallAttemptList.elements.toList
+      singleCallAttemptCost = computeCost(singleCallName, singleCallAttempt)
```
Comment (THWiseman) [inline services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:610]: What is your current thinking about how we should handle calls with multiple attempts?

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala [hunk 5]
```diff
@@ -488,6 +515,134 @@ class MetadataBuilderActor(readMetadataWorkerMaker: () => Props,
...
+          val startTimeErrorOr = ErrorOr(OffsetDateTime.parse(startTimeVal.value))
+
+          val endTimeErrorOr =
+            extractFromJsAs[JsString](jsCall, CallMetadataKeys.VmEndTime)
+              .map(v => ErrorOr(OffsetDateTime.parse(v.value)))
+              .getOrElse(OffsetDateTime.now().validNel)
+
+          val rawCostPerHour = extractFromJsAs[JsNumber](jsCall, CallMetadataKeys.VmCostPerHour).map(_.value)
+          val costPerHourErrorOr = rawCostPerHour match {
+            case Some(c: BigDecimal) if c >= 0 => c.validNel
```
Comment (THWiseman) [inline services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:571]: TIL that `JsNumber` and `BigDecimal` are the same type under the covers. Was surprised that this match works!

https://www.javadoc.io/doc/io.spray/spray-json_2.9.3/latest/spray/json/JsNumber.html

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala (hunk 3 above)
Comment (jgainerdewar) [reply services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:589]: I was wondering whether GCP publishes what unit they use to do their pricing calculations, have you ever come across that? Agree that seconds is a good default if not.

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala (hunk 4 above)
Comment (jgainerdewar) [reply services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:610]: I think they should be handled correctly here, it should sum the costs for all of the attempts.

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala (hunk 3 above)
Comment (THWiseman) [reply services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:589]: They sure do. It's different for different things. You can see an example [here](https://cloud.google.com/billing/v1/how-tos/catalog-api#getting_the_list_of_skus_for_a_service). 

They mostly use cost per hour. "cost" is USD/hr, but the USD are measured down to the nano dollar. Timestamps are accurate to the nanosecond.

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala [hunk 6]
```diff
@@ -488,6 +567,94 @@ class MetadataBuilderActor(readMetadataWorkerMaker: () => Props,
...
+      case _ => JsArray.empty.elements
+    }
+
+    val resp = JsObject(
+      Map(
+        WorkflowMetadataKeys.Id -> JsString(id.toString),
+        WorkflowMetadataKeys.Status -> JsString(status),
+        "currency" -> JsString(DefaultCurrency.getCurrencyCode),
+        "cost" -> JsNumber(callCost + subworkflowCost),
+        "errors" -> JsArray(costErrors ++ subworkflowErrors)
```
Comment (salonishah11) [inline services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:650]: If there are any errors when computing cost, should the `cost` field be 0? Or is the intention for it to always reflect the approximate cost of jobs that were successfully calculated?

**Review by salonishah11 — APPROVED** [review]

File: services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala (hunk 6 above)
Comment (jgainerdewar) [reply services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala:650]: The latter - then the caller can decide how much they want to believe that cost. My thinking is that if it's a huge workflow and we failed to compute the cost for just one task, it's still useful to know the cost of the rest of it.

**Review by THWiseman — APPROVED** [review]

**Pull request merged**